### PR TITLE
Frogger

### PR DESCRIPTION
**Pull request by AnnaStoyano**
### Frogger

[Demo](https://annastoyano.github.io/Frogger/index) |
[Code base](https://github.com/AnnaStoyano/Frogger)

Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
@AnnaStoyano looks nice! :)

While we're looking for some review capacity please feel free checking your code against the following list of requirements.

## Check-list

Minimal requirements to meet:
 * [ ] it is OK to employ ES6 features like `const`, `let` etc.
 * [ ] OO is implemented using JS prototype chain object model (**not** ES6 classes syntax)
 * Requirements re **Constants**:
   * [ ] all numbers like block dimensions, initial locations are defined as constants
   * [ ] there are core constants and derived constants
         (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
   * [ ] arrays of constants are also constants 
         (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
   * [ ] const objects help organizing and structure const data even better
         (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
 * Requirements re **OOP**:
   * [ ] properties common for some classes are generalized into a base class
         (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
   * [ ] class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
         not `Subclass.prototype = new Superclass(params);`;
         [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
   * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
         (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
         Hint: pass `Player` instance as an argument to every enemy
   * [ ] Separation of Concerns principle is followed
         (e.g. `update` method does only rendering and 
         doesn't **contain** any inline code to check e.g. collisions)
   * [ ] the code is very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

File: submissions/AnnaStoyano/frogger/app.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+Enemy.prototype = Object.create(Entity.prototype);
+Enemy.prototype.update = function (dt) {
+    this.position.x += this.speed * dt * (this.level * 0.8);
+    this.checkWallCollision(canvas.width, canvas.height, -90, 0);
+    this.checkLose();
+}
+
+Enemy.prototype.checkCollision = function () {
+    let positionX = Math.floor(this.position.x);
+    if (positionX - 50 < player.position.x && player.position.x < positionX + 50) {
```
Comment (OleksiyRudenko) [inline submissions/AnnaStoyano/frogger/app.js:48]: Here `player` is a global variable.

From the checklist:
> classes do not refer to any global variables, like global variable player, which is an instance of Player class
(referring to global constants and globals provided by the gaming platform like Resources is OK);

Double check your entire code base to meet the requirement.

File: submissions/AnnaStoyano/frogger/app.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            this.position.y += 20;
+            break;
+    }
+}
+
+const player = new Player(200, 404, 7, 'images/char-boy.png');
+
+const enemy1 = new Enemy(30, 120, 60, 'images/enemy-bug.png', player);
+const enemy2 = new Enemy(15, 220, 30, 'images/enemy-bug.png', player);
+const enemy3 = new Enemy(0, 50, 20, 'images/enemy-bug.png', player);
```
Comment (OleksiyRudenko) [inline submissions/AnnaStoyano/frogger/app.js:120]: 200, 404, 7, 220 etc look look magic

From the requirements:
> all numbers like block dimensions, initial locations are defined as constants

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@AnnaStoyano thank you for the update.

The below comments comprise a partial review.

Please check your code against the checklist and:
> Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by AnnaStoyano** [comment]
@OleksiyRudenko  Thank you for your comments. I hope that I corrected my mistakes.

File: submissions/AnnaStoyano/frogger/app.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+const ENEMY_1_SPEED = 20;
+const ENEMY_2_SPEED = 60;
+const ENEMY_3_SPEED = 30;
+
+
+const player = new Player(START_POSITION_PLAYER.X, START_POSITION_PLAYER.Y, PLAYER_SPEED, 'images/char-boy.png');
+const enemy1 = new Enemy(START_POSITION_ENEMY_1.X, START_POSITION_ENEMY_1.Y, ENEMY_1_SPEED, 'images/enemy-bug.png', player);
+const enemy2 = new Enemy(START_POSITION_ENEMY_2.X, START_POSITION_ENEMY_2.Y, ENEMY_2_SPEED, 'images/enemy-bug.png', player);
+const enemy3 = new Enemy(START_POSITION_ENEMY_3.X, START_POSITION_ENEMY_3.Y, ENEMY_3_SPEED, 'images/enemy-bug.png', player);
+const allEnemies = [enemy1, enemy2, enemy3];
```
Comment (OleksiyRudenko) [inline submissions/AnnaStoyano/frogger/app.js:133]: Whenever you feel adding an index in a variable name (like `ENEMY_1_SPEED`, `START_POSITION_ENEMY_3` or `enemy2`) this is clear sign that array would be more appropriate.
Imagine you'd be required to add values for dozens of characters.
There is also massive code repetition here (again, think of dozens of characters).
Can you devise an approach that would employ data structure (array of objects containing individual enemy settings)?
You might then efficiently use e.g. `Array.map`.

Hint:
`allEnemies = enemySettings.map(settings => new Enemy(settings.startPosition.X, settings.startPosition.Y, settings.speed, ENEMY_IMAGE, player))`

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@AnnaStoyano well done. Yet you will benefit more if you master data structures to save time and lines of code.

**Comment by AnnaStoyano** [comment]
@OleksiyRudenko  Thank you for your advice and hint)

**Review by OleksiyRudenko — APPROVED** [review]
@AnnaStoyano well done!
There are still numbers with unclear semantic purpose worth being assigned in constant definitions.
At all times keep thinking of how your code **reads**. The closer to plain English the better.

**Pull request merged**